ESQL: Log start and end of queries

alex-spies opened the pull request:
To make sure that ESQL queries that do not succeed can always be identified in the logs, log queries both before and after execution. ~~To correlate starts and ends in case of concurrent queries, generate a query ID and use it in the logs.~~ 
**Update:** Since this adds undue complexity, this PR does not add any kind of ID to queries anymore.

The generated logs look like this:
```
[2023-09-21T12:15:28,213][INFO ][o.e.x.e.a.RestEsqlQueryAction] [runTask-0] Beginning execution of ESQL query.
Query string: [row a = 1 | eval x = now() + 1year]
[2023-09-21T12:15:28,286][INFO ][o.e.x.e.a.RestEsqlQueryAction] [runTask-0] Finished execution of ESQL query.
Query string: [row a = 1 | eval x = now() + 1year]
Execution time: [72]ms
```


# Generated query ID vs. session ID

~~Some people suggested logging the [session ID](https://github.com/elastic/elasticsearch/blob/d0e91c9bef145e5ceee9df66603baf4639d0cba4/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java#L96). This is actually tricky because the session ID is created _after_ the `EsqlResponseListener` [is created ](https://github.com/elastic/elasticsearch/blob/d0e91c9bef145e5ceee9df66603baf4639d0cba4/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RestEsqlQueryAction.java#L53) in `RestEsqlQueryAction::prepareRequest`.~~

~~To create a log message that contains the execution time _and_ the session ID, the session ID would need to be passed to the `EsqlResponseListener` and the only sensible way I found was attaching the session ID to the `EsqlQueryResponse` that is passed to the listener; changing the `EsqlQueryResponse` would blow up the scope here just for the sake of logging esp. since the query response needs to be serializable.~~

~~Another way to have a log that correlates an incoming ESQL query with a session ID would be generating a query ID and [passing it along with the `EsqlQueryRequest` ](https://github.com/elastic/elasticsearch/blob/d0e91c9bef145e5ceee9df66603baf4639d0cba4/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RestEsqlQueryAction.java#L52) so that, once the session ID is generated, we can log something like `Starting compute session with ID [...] for query ID [...]`. The scope increase is more managable this way, but we would litter the code with query logging in multiple places, and would still have to consider serializability issues.~~

~~This is why I think that just generating an arbitrary query ID is more pragmatic. Moreover, for a future async API, a request ID will have to be generated and used anyway, which then can replace the query ID that is introduced here.~~

alex-spies committed:
Correlate start and end of ESQL queries in logs

alex-spies committed:
Rename variable for readability

alex-spies committed:
Structure logs consistently

alex-spies committed:
Move query logging code into RestEsqlQueryAction

To correlate start and end of queries more easily, move all query
logging code into RestEsqlQueryAction. This way, the start and end logs
of a query will show up logged under the same class, and the logging
code is nicely in one place.

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

alex-spies committed:
Update docs/changelog/99746.yaml

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies committed:
Remove obsolete logger

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-query-logging-improvements

luigidellaquila commented:
Thanks Alex, I had a quick look at the PR and I'm not sure having a query ID justifies all this complexity (refactoring the stopwatch, the additional wrapping...)
Let's discuss it off-line and see if we can simplify it a bit

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ThreadSafeStopWatch.java:
```diff
@@ -0,0 +1,53 @@
+ * A simple, thread-safe stop watch for timing a single action.
+ * Allows to stop the time for building a response and to log it at a later point.
+ */
+class ThreadSafeStopWatch {
```
No need for this class - it's used in only one place where the thread-safety is not needed.

alex-spies replied:
This PR only moves this class to a new place. It was added based on feedback that we should log the same time that the user also sees, which requires a state change (stopping the time when the query response was built), which raised (minor) thread safety concerns (I guess `onResponse` and `onFailure` could be triggered concurrently, but both need to access the stopwatch).

If it's okay, I'd like not to drop this class in this PR as it would affect how we measure execution time of successful queries.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RestEsqlQueryAction.java:
```diff
@@ -40,21 +45,64 @@ public List<Route> routes() {
             esqlRequest = EsqlQueryRequest.fromXContent(parser);
         }
+        // Create some kind of query id so that we can correlate the beginning and the end of a query in the logs.
+        String queryId = UUIDs.base64UUID();
```
What's the point of the id? to differentiate between queries or is it meant to be persisted somewhere?
As far as I can tell it's the former and thus needs no time component so doing queryId.hashCode() should work - it will be the same for the exact same queries but that is fine.

alex-spies replied:
The point of the ID is to be able to correlate start and end of a query, even if e.g. the same query string gets sent multiple times (user hits send twice or Kibana sends the query multiple times etc.). Mostly, I wanted to address a suggestion to log the session ID, but using something simpler because logging the session ID is actually very non-trivial.

I'll just remove it if it's not useful enough. A hash code wouldn't help since the id should be different even for the same query being sent twice.

costin reviewed (changes requested):
There's too much activity around a small issue; no need for a stopwatch, worry about thread-safety (only local variables that are not written to and that exist inside a listener).
The listener wrapping is there - unless there's an actual need for the UUID (which makes the message longer), just use the hashcode and call it a day.
It's why I'm marking this PR as requesting changes - no need to have a lot of code for a simple issue.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -31,72 +26,36 @@
  * Listens for a single {@link EsqlQueryResponse}, builds a corresponding {@link RestResponse} and sends it.
  */
 public class EsqlResponseListener extends RestResponseListener<EsqlQueryResponse> {
```
This class has no intrinsic value and its content moved over to the RestEsqlQueryAction.

alex-spies replied:
To clarify: do you mean `EsqlResponseListener`? If so, I agree that it could probably also be a nested (private static) class inside `RestEsqlQueryAction`; however, this already existed before any recent logging PRs so refactoring this would be out of scope for this PR IMO.

alex-spies committed:
Remove query ID

alex-spies committed:
Simplify

- Move ThreadSafeStopWatch back to where it was initially.
- Perform logging after queries inside EsqlResponseListener again.

alex-spies commented:
@costin : I simplified the PR, the diff should now be minimal for what we try to achieve. No more query ID and no moving around of code (`ThreadSafeStopWatch` and `wrapWithLogging` are in their original places again).

I would like to follow this up with a refactoring that moves the logging of start and end of queries both into `RestEsqlQueryAction`; this is relevant because the start and end logs should really be logged under the same class to avoid confusion when people search these logs later. The refactoring should then be fast to review, though, even though it may require to move around the `ThreadSafeStopWatch` after all.

costin reviewed (approved):
LGTM

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -143,12 +143,16 @@ public ActionListener<EsqlQueryResponse> wrapWithLogging() {
             // At this point, the StopWatch should already have been stopped, so we log a consistent time.
-            LOGGER.info("Successfully executed ESQL query in [{}]ms: [{}]", stopWatch.stop().getMillis(), esqlQuery);
+            LOGGER.info(
+                "Finished execution of ESQL query.\nQuery string: [{}]\nExecution time: [{}]ms",
```
@alex-spies I've noticed these messages today while testing stuff and I was wondering why they look like `Failed executing ESQL query in [8]ms: [show functions | where variadic is true | keep synopsis]`. I understand the square brackets surrounding the query itself (to make it stand out and be clearly readable), but I don't get why we have `[8]ms`? I don't think they are necessary, are they?

alex-spies replied:
Yeah, `[8]ms` looks a bit irritating to myself as well. However, this is consistent with how our [contribution instructions ](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#logging)say that we should log, there's even exactly this example where they also write `... failed in [{}]ms`.

astefan replied:
Imho that's a mistake. I searched ES code for both `[{}]ms` and `[{}ms]` and the second approach is much more common.
